src/type_sys/fender_value/fender_strings.rs

```diff
@@ -0,0 +1,247 @@
+
+impl FenderString {
+    pub fn escaped_string(&self) -> String {
+        format!("{:?}", self.to_string())
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.chars.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.chars.len()
+    }
+
+    pub fn push_str(&mut self, other: &str) {
+        if other.is_empty() {
+            return;
+        }
+
+        self.chars.extend(other.chars());
+
+        if let Some(s) = self.cached_string.borrow_mut().as_mut() {
+            s.push_str(other);
+        }
+    }
+
+    pub fn push_fndr_str(&mut self, other: &FenderString) {
+        if other.is_empty() {
+            return;
+        }
+
+        self.chars.extend(other.chars.clone());
+
+        if let Some(s) = self.cached_string.borrow_mut().as_mut() {
+            for c in other.iter() {
+                s.push(*c);
+            }
+        }
+    }
+
+    pub fn push_char(&mut self, other: char) {
+        self.chars.push(other);
+
+        if let Some(s) = self.cached_string.borrow_mut().as_mut() {
+            s.push(other);
+        }
+    }
+
+    pub fn remove(&mut self, pos: usize) -> Option<char> {
+        if pos >= self.len() {
+            None
+        } else {
+            if self.cacheable {
+                *self.cached_string.get_mut() = None;
+            }
+            Some(self.chars.remove(pos))
+        }
+    }
+
+    pub fn remove_at(&mut self, mut pos: i64) -> Result<char, String> {
+        pos = if pos < 0 {
+            self.len() as i64 + pos
+        } else {
+            pos
+        };
+
+        let pos = if pos < 0 {
+            return Err("Invalid wrapping index".into());
+        } else {
+            pos as usize
+        };
+
+        if pos >= self.len() {
+            Err(index_oob!(pos, self.len()))
+        } else {
+            if self.cacheable {
+                *self.cached_string.get_mut() = None;
+            }
+            Ok(self.chars.remove(pos))
+        }
+    }
+
+    pub fn to_uppercase(&self) -> FenderString {
+        FenderString::new(
+            self.chars
+                .iter()
+                .map(|c| c.to_uppercase().next().unwrap())
+                .collect(),
+        )
+    }
+    pub fn to_lowercase(&self) -> FenderString {
+        FenderString::new(
+            self.chars
+                .iter()
+                .map(|c| c.to_lowercase().next().unwrap())
+                .collect(),
+        )
+    }
+
+    pub fn iter(&self) -> std::slice::Iter<'_, char> {
+        self.chars.iter()
+    }
+
+    pub fn chars(&self) -> &Vec<char> {
+        &self.chars
+    }
+
+    pub fn swap(&mut self, pos_a: usize, pos_b: usize) -> bool {
+        if pos_a >= self.len() || pos_b >= self.len() {
+            return false;
+        }
+        if self.cacheable {
+            *self.cached_string.get_mut() = None;
+        }
+        self.chars.swap(pos_a, pos_b);
+        true
+    }
+
+    pub fn insert_char(&mut self, index: usize, value: char) -> bool {
```

Comment:
With us introducing all of these functions, I would like to have at least a few unit tests for them. Could we create something like `ensure_valid` that makes sure the char vec and the cached string are identical, and then use that for assertions in tests?